### PR TITLE
WL-1242: Specify Data Sharing Consent wording on a per Enterprise basis

### PR DESCRIPTION
**Pull request by ziafazal**
**Description:** This PR has changes to specify Data Sharing Consent wording on a per Enterprise basis.

**JIRA:** https://openedx.atlassian.net/browse/WL-1242

**Dependencies:** None

**Merge deadline:** None

**Testing instructions:**

1. Install edx-enterprise in develop mode and checkout ziafazal/WL-1512 branch
2. Enable edx-enterprise in edx-platform via feature flag
3. Run migrations
`./manage lms migrate --settings=devstack`
4. Add New enterprise customer via django admin
Home › Enterprise › Enterprise Customers
Make sure `Enable data sharing consent` checkbox is checked and `Enforce data sharing consent` is set to `At Enrollment`
5. Create a Data sharing consent page for newly created enterprise customer via django admin
Home › Enterprise Consent › Data sharing consent pages › Add data sharing consent page
6. After saving Data sharing consent page open it again in edit mode. Click on the `Preview (course)` button or `Preview (program)` on top right to preview your newly created data sharing page in course or program context.
7. Repeat step 4, 5 and 6 to create Data sharing consent page for an other customer and preview it.

**Sandbox Testing instructions:**
The latest version of this branch is installed at:

- https://bill-filler.sandbox.edx.org
- edx-themes has been updated to remove the MIT overrides for DSC files
- 3 enterprise customers are configured (EdXEnterprise, Boeing, Haliburton), this can be seen via django-admin https://bill-filler.sandbox.edx.org/admin/enterprise/enterprisecustomer/
- Boeing and Haliburton are each associated with site https://mitxpro-bill-filler.sandbox.edx.org. EdxEnterprise is associated with https://bill-filler.sandbox.edx.org
- Custom Data Sharing Consent pages have been setup for Boeing and Haliburton https://bill-filler.sandbox.edx.org/admin/consent/datasharingconsentpage/

1. Add a separate learner for each enterprise via Enterprise->Enterprise Customer->Manage Learners
2. For EdxEnterprise, manually enroll learner edx@test.com in course course-v1:edX+DemoX+Demo_Course
3. For Boeing, manually enroll learner boeing@test.com in course course-v1:MITxPRO+WL_E2E+2018
4. For Halliburton, manually enroll learner hal@test.com in course course-v1:MITxPRO+WL_E2E+2018
5. For EdXEnterprise, go to bill-filler.sandbox.edx.org and create an account for edx@test.com. Then go to dashboard and try to access course-v1:edX+DemoX+Demo_Course. You should see the default Data Sharing Consent form
6. For Boeing, got to mitxpro-bill-filler.sandbox.edx.org, and create an account for boeing@test.com
7. Then go to dashboard and try to access course-v1:MITxPRO+WL_E2E+2018. You should see customized Data Sharing Consent form configured for Boeing.
8. For Halliburton, got to mitxpro-bill-filler.sandbox.edx.org, and create an account for hal@test.com
9. Then go to dashboard and try to access course-v1:MITxPRO+WL_E2E+2018. You should see customized Data Sharing Consent form configured for Halliburton.

**Merge checklist:**

- [x] New requirements are in the right place (`base.in` if only used in enterprise; in the correct `platform-****.in` files if they're hosted in edx-platform)
- [x] All reviewers approved
- [x] CI build is green
- [x] Version bumped
- [x] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [x] Commits are (reasonably) squashed
- [ ] Translations are updated
- [x] PR author is listed in AUTHORS

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is finished.
- [ ] Delete working branch (if not needed anymore)
- [ ] edx-platform PR (be sure to include edx-platform requirements upgrades that were present in this PR)

**Author concerns:** List any concerns about this PR - inelegant 
solutions, hacks, quick-and-dirty implementations, concerns about 
migrations, etc.

### CODE REVIEW

**Comment by bill-filler** [comment]
@ziafazal 
It's working really nicely, great job!
Some comments to follow.
First one: we should not be changing all of the requirements/* files with new versions of everything. Should just be one changed requirements file to add bleach

File: consent/models.py [hunk 1]
```diff
@@ -239,3 +245,151 @@ class Meta(Consent.Meta):
...
+        '<ul>'
+        '<li>enterprise_customer_name: A name of enterprise customer.</li>'
+        '<li>platform_name: Name of platform.</li>'
+        '</ul>'
+    ))
+    SIDEBAR_PARAGRAPH_HELP_TEXT = mark_safe_lazy(_(
+        'Fill in a text for left sidebar paragraph. The following variables may be available:<br />'
+        '<ul>'
+        '<li>enterprise_customer_name: A name of enterprise customer.</li>'
+        '<li>platform_name: Name of platform.</li>'
```
Comment (bill-filler) [inline consent/models.py]: add item: help text

File: consent/models.py [hunk 2]
```diff
@@ -239,3 +245,151 @@ class Meta(Consent.Meta):
...
+        '</ul>'
+    ))
+    POLICY_PARAGRAPH_HELP_TEXT = mark_safe_lazy(_(
+        'Fill in a text for policy paragraph at the bottom of page. The following variables may be available:<br />'
+        '<ul>'
+        '<li>enterprise_customer_name: A name of enterprise customer.</li>'
+        '<li>platform_name: Name of platform.</li>'
```
Comment (bill-filler) [inline consent/models.py]: add item: help text

File: consent/models.py [hunk 3]
```diff
@@ -239,3 +245,151 @@ class Meta(Consent.Meta):
...
+    """
+
+    class Meta(object):
+        app_label = 'consent'
+
+    TOP_PARAGRAPH_HELP_TEXT = mark_safe_lazy(_(
+        'Fill in a text for first paragraph of page. The following variables may be available:<br />'
+        '<ul>'
+        '<li>enterprise_customer_name: A name of enterprise customer.</li>'
+        '<li>platform_name: Name of platform.</li>'
```
Comment (bill-filler) [inline consent/models.py]: add item: help text

File: consent/models.py [hunk 4]
```diff
@@ -239,3 +245,151 @@ class Meta(Consent.Meta):
...
+        default=_('Data sharing consent required'),
+        help_text=_('Title of page')
+    )
+    left_sidebar_text = models.TextField(
+        null=True,
+        blank=True,
+        default=_(
+            "<p>Welcome to {platform_name}</p>"
+            "<p>{enterprise_customer_name} has partnered with {platform_name} to offer you high-quality learning "
+            "opportunities from the world's best universities.</p>"
```
Comment (bill-filler) [inline consent/models.py]: To try and match default styles, change to this:
```
<p class="partnered-text">Welcome to {platform_name}</p>
<p class="partnered-text"><strong>{enterprise_customer_name}</strong> has partnered with <strong>{platform_name}</strong> to offer you high-quality learning opportunities from the world's best universities.</p>
```

File: consent/models.py [hunk 5]
```diff
@@ -239,3 +245,151 @@ class Meta(Consent.Meta):
...
+    )
+    top_paragraph = models.TextField(
+        null=True,
+        blank=True,
+        default=_('<h2>Consent to share your data</h2>'
+                  '<p>To access this {item}, you must first consent to share your learning achievements with '
+                  '<b>{enterprise_customer_name}</b></p>'
+                  '<p>{enterprise_customer_name} would like to know about:</p>'
+                  '<ul><li>your enrollment in this course</li><li>your learning progress</li>'
+                  '<li>course completion</li></ul>'),
```
Comment (bill-filler) [inline consent/models.py]: Same as above, try to match default style:
```
<h2 class="consent-title">Consent to share your data</h2>
<p>To access this {item}, you must first consent to share your learning achievements with 
<b>{enterprise_customer_name}</b></p>
<p>{enterprise_customer_name} would like to know about:</p>
<ul><li>your enrollment in this course</li><li>your learning progress</li>
<li>course completion</li></ul><p>
```
make sure to add trailing paragraph tag

**Comment by bill-filler** [comment]
I will do another round of testing and review next week, ran out of time today

**Comment by bill-filler** [comment]
Need to get the tests passing, hopefully merging https://github.com/edx/edx-enterprise/pull/298 will make them pass?

**Comment by douglashall** [comment]
Looks like the version of bleach installed with edxapp is [1.4](https://github.com/edx/edx-platform/blob/master/requirements/edx/base.txt#L11). It would be best to stick to that version if possible. I recommend adding the bleach requirement manually to both the .in and .txt files for named releases/master and not running pip-compile.

**Comment by ziafazal** [comment]
@bill-filler  @douglashall I have changed bleach version to 1.4.

**Comment by bill-filler** [comment]
@ziafazal 
When previewing for program mode, the value of {item} always is being displayed as "course". I expected it to be "program"

**Comment by bill-filler** [comment]
@ziafazal 
please make sure {course_title} and {course_start_date} are available as variables for the consent page as we'll need these in order to remove the overriden template in the mitxpro theme: https://github.com/edx/edx-themes/blob/master/edx-platform/mitxpro.mit.edu/lms/templates/enterprise/_data_sharing_policy.html#L8

**Comment by bill-filler** [comment]
@ziafazal 
I just realized we also override the template that gets displayed in the banner of the dashboard when a user declines consent: https://github.com/edx/edx-themes/blob/master/edx-platform/mitxpro.mit.edu/lms/templates/enterprise_support/enterprise_consent_declined_notification.html

We should probably add two additional fields to the model:
declined_notification_title: make sure course_name/course_title are available as variables
declined_notification_message: the message to display

And then change the default template to support these fields.
Here is a screenshot of what it looks like with MIT override:
declinednotification

**Comment by bill-filler** [comment]
@douglashall 
Would you mind giving this an initial review? It's configured and can be tested from my sandbox (see Sandbox Testing Instructions) in description.

**Comment by bill-filler** [comment]
@ziafazal @ibrahimahmed443 
All looking good after last changes

File: consent/models.py [hunk 6]
```diff
@@ -239,3 +245,193 @@ class Meta(Consent.Meta):
...
+    ))
+    NOTIFICATION_MESSAGE_HELP_TEXT = mark_safe_lazy(_(
+        'Fill in a text for message of the notification which appears on dashboard '
+        'when user decline to provide consent. '
+        'The following variables may be available:<br />'
+        '<ul>'
+        '<li>enterprise_customer_name: A name of enterprise customer.</li>'
+        '<li>course_title: Title of course. Available when type of consent is course.</li>'
+        '</ul>'
+    ))
```
Comment (douglashall) [inline consent/models.py]: You should chunk up the `_data_sharing_policy.html` template into smaller template partials that match the chunks that you want to be customizable per EnterpriseCustomer. Make the fields in this model optional with no defaults. Then in each of the template partials you created by the chunking about, check the model to see if the chunk has an override.

File: consent/models.py [hunk 7]
```diff
@@ -239,3 +245,193 @@ class Meta(Consent.Meta):
         help_text=_("Course key for which data sharing consent is granted.")
     )
     history = HistoricalRecords()
+
+
+@python_2_unicode_compatible
+class DataSharingConsentPage(TimeStampedModel):
```
Comment (douglashall) [inline consent/models.py]: I would call this something like `DataSharingConsentTextOverrides`.

File: consent/models.py (hunk 6 above)
Comment (ziafazal) [reply consent/models.py]: we opted to go with default values or override approach to 
1) give ability to skip any text/sections. i.e if an enterprise customer don't want to show policy dropdown they can set its values to empty.
2) defaults give users an idea of typical html for certain section

**Comment by bill-filler** [comment]
@douglashall 
To implement the MIT design, we actually need to expose all of the fields that we currently expose in DataSharingConsentTextOverrides as they all need to be overriden to implement their desired design.

File: consent/models.py (hunk 6 above)
Comment (bill-filler) [reply consent/models.py]: @douglashall 
for the reasons @ziafazal pointed out, are you ok with going with the approach we have implemented?

Comment (douglashall) [reply consent/models.py]: No. I really don't like defining this HTML as strings in python code. It should be defined in a template file where an engineer would expect HTML code to be defined. You can always define an override with an empty string value if that is what the desired override needs to be.

**Comment by douglashall** [comment]
@bill-filler Is there any consolidation of the fields that can be done so there are not so many overrides?

File: consent/models.py (hunk 6 above)
Comment (ziafazal) [reply consent/models.py]: @douglashall these HTML strings are going to be displayed only in django admin where staff can override those according to their needs. Should i override django admin template? I'm reluctant to do that for maintainability reasons.

Comment (douglashall) [reply consent/models.py]: Yes, I have less of an issue with the help strings you are defining here. The default strings you have defined on the fields below are still a problem.

**Comment by bill-filler** [comment]
@ziafazal 
couple of issues during testing:

1. Text title incorrect in django-admin it says
```
Data sharing consent text overridess
```
Should say 
```
Data sharing consent text overrides
```
2. I'm not getting any default value in django-admin when I create a new DSC text override

**Comment by ziafazal** [comment]
@bill-filler I have fixed no.1 about no.2 purpose of those default values was only to show them in django admin. Removing them from `models.py` means no defaults on django admin.

File: consent/models.py [hunk 8]
```diff
@@ -239,3 +245,159 @@ class Meta(Consent.Meta):
...
+        related_name="data_sharing_consent_page",
+        on_delete=models.deletion.CASCADE
+    )
+    published = models.BooleanField(
+        default=False,
+        help_text=_(
+            _("Specifies whether data sharing consent page is published.")
+        )
+    )
+    history = HistoricalRecords()
```
Comment (douglashall) [inline consent/models.py]: We are trying to remove all uses of `django-simple-history`. Please remove it from this model.

**Comment by ziafazal** [comment]
@douglashall removed `django-simple-history` usage in `DataSharingConsentTextOverrides` model.

**Pull request merged**